allennlp/modules/elmo.py

```diff
@@ -681,3 +685,148 @@ def create_cached_cnn_embeddings(self, tokens: List[str]) -> None:
+
+        return fc_layer
+
+    def _chunked_log_probs(self,
+                           activation: torch.Tensor) -> torch.Tensor:
+        """
+        Do the softmax in chunks so the gpu ram doesn't explode.
+
+        Parameters
+        ----------
+        activation : ``torch.Tensor``, required
+            Results of the ELMo embedder.
+
+        Returns
+        -------
+        ``torch.Tensor`` with all of the log probabilities.
+        """
+        all_logprobs = []
+        num_chunks = (activation.size(0) - 1) // self.chunk_size + 1
+        for activation_chunk in torch.chunk(activation, num_chunks, dim=0):
+            chunk_size, sequence_length = activation_chunk.size()[:2]
+
+            time_indexer = (torch.arange(0, chunk_size * sequence_length)
+                            % sequence_length)
+
+            batch_indexer = (torch.arange(0, chunk_size * sequence_length)
+                             / sequence_length)
+
+            all_logprobs.append(
+                F.log_softmax(self.fc_layer(activation_chunk), 2)[
+                batch_indexer, time_indexer, :].view(
+                    *[chunk_size, sequence_length, -1]))
+
+        return torch.cat(all_logprobs, 0)
+
+    def forward(self,
+                bilm_inputs: Dict[str, Union[torch.Tensor, List[torch.Tensor]]],
+                aggregation_function: str = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Passes the ELMo embeddings to a softmax layer and calculates
+        the log probabilities.
+
+        Parameters
+        ----------
+        bilm_inputs : ``Dict[str, Union[torch.Tensor, List[torch.Tensor]]]``, required
+            The output from `_ElmoBiLm`.
+        word_inputs : ``torch.Tensor``, required
+            The output from `batch_to_ids` when `
+        aggregation_function : ``str``, optional
+            An aggregation function that aggregates the forward and backward
+            tensors. Default is `None` where no aggregation is done.
+
+        Returns
+        -------
+        A tuple of `log_probs` and `mask_without_bos_eos`.
+        """
+
+        activations = bilm_inputs['activations']
+        mask = bilm_inputs['mask']
+
+        final_activations = activations[-1]  # (batch, BOS + L + EOS, [fwd, bwd])
+
+        # get rid of predicting, or conditioning on EOS for the forward
+        # activation, and BOS for the reverse activation.
+        mask_without_bos_eos = mask[:, 2:]
+
+        # NOTE: can this be replaced by allennlp.util.masked_softmax ?
+        fwd_log_probs = self._chunked_log_probs(
+            final_activations[:, :-2, :self.hidden_size])
+        bwd_log_probs = self._chunked_log_probs(
+            final_activations[:, 2:, self.hidden_size:])
+
+        # [B, (fwd, bwd), L]
+        log_probs = torch.stack((fwd_log_probs, bwd_log_probs), 1)
+        log_probs *= mask_without_bos_eos[:, None, :, None].float()
+
+        # aggregate, otherwise return both the backward and forward embeddings
```

Comment:
Remove these aggregation functions and `aggregation_function` from the `forward` arguments and leave that as a post-processing step.  Just return a dict: `{'forward_log_probs':  `fwd_log_probs`, 'backward_log_probs':  `bwd_log_probs`, `mask`: `mask_without_bos_eos`}`.

This simplifies the code here (and the calculation of both the `mean` and `max` below is incorrect anyway).